File: TsinghuaNet.Helper/Models/Command.vb
Imports System.ComponentModel
Imports System.Windows.Input

Public Class Command
    Implements ICommand

    Private viewModel As NetObservableBase
    Private action As Action(Of Object)
    Public Sub New(viewModel As NetObservableBase, action As Action)
        Me.New(viewModel, Sub(o) action())
    End Sub
    Public Sub New(viewModel As NetObservableBase, action As Action(Of Object))
        Me.viewModel = viewModel
        Me.action = action
        AddHandler viewModel.PropertyChanged, AddressOf OnModelPropertyChanged
    End Sub

    Private Sub OnModelPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
        If e.PropertyName = "IsBusy" Then
            OnCanExecuteChanged()
        End If
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
    Protected Overridable Sub OnCanExecuteChanged()
        RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
    End Sub

    Public Sub Execute(parameter As Object) Implements ICommand.Execute
        Try
            viewModel.IsBusy = True
            action?.Invoke(parameter)
        Finally
            viewModel.IsBusy = False
        End Try
    End Sub

    Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return Not viewModel.IsBusy
    End Function
End Class

File: TsinghuaNet/Models/Command.cs
using System;
using System.ComponentModel;
using System.Windows.Input;

namespace TsinghuaNet.Models
{
    public class Command : ICommand
    {
        private NetViewModelBase viewModel;
        private Action<object> action;
        public Command(NetViewModelBase viewModel, Action action) : this(viewModel, o => action())
        {
        }
        public Command(NetViewModelBase viewModel, Action<object> action)
        {
            this.viewModel = viewModel;
            this.action = action;
            viewModel.PropertyChanged += OnModelPropertyChanged;
        }

        private void OnModelPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "IsBusy")
                OnCanExecuteChanged();
        }

        public event EventHandler CanExecuteChanged;
        protected virtual void OnCanExecuteChanged() => CanExecuteChanged?.Invoke(this, EventArgs.Empty);

        public void Execute(object parameter)
        {
            try
            {
                viewModel.IsBusy = true;
                action?.Invoke(parameter);
            }
            finally
            {
                viewModel.IsBusy = false;
            }
        }

        public bool CanExecute(object parameter) => !viewModel.IsBusy;
    }
}
